clustercache/clustercache.go

```diff
@@ -0,0 +1,240 @@
+package clustercache
+
+import (
+	"fmt"
+	"path"
+	"time"
+
+	"github.com/fatih/color"
+	"github.com/giantswarm/columnize"
+	"github.com/giantswarm/gscliauth/config"
+	"github.com/giantswarm/gsclientgen/models"
+	"github.com/giantswarm/gsctl/client"
+	"github.com/giantswarm/gsctl/client/clienterror"
+	"github.com/giantswarm/gsctl/commands/errors"
+	"github.com/giantswarm/gsctl/confirm"
+	"github.com/giantswarm/microerror"
+	"github.com/spf13/afero"
+	yaml "gopkg.in/yaml.v2"
+)
+
+const (
+	listClustersActivityName = "list-clusters"
+	clusterCacheFileName     = "clustercache.yaml"
+
+	cacheDuration = time.Hour * 24 * 7 // 7 days
+	timeLayout    = time.RFC3339
+)
+
+// EndpointCache stores the IDs stored in an
+// endpoint-specific cache, and also its expiry date
+type EndpointCache struct {
+	Expiry string   `yaml:"expiry"`
+	IDs    []string `yaml:"ids"`
+}
+
+// Endpoints stores a map with the keys being API endpoints,
+// and the values being caches
+type Endpoints map[string]EndpointCache
+
+// Cache is the file structure of the cluster cache file
+type Cache struct {
+	Endpoints Endpoints `yaml:"endpoints"`
+}
+
+// GetID gets the cluster ID for a provided name/ID
+// by checking in both the user cache and on the API
+func GetID(endpoint string, clusterNameOrID string, clientWrapper *client.Wrapper) (string, error) {
+	// Check if the cluster ID is already in the cache,
+	// and skip the API request if it is
+	isInCache := IsInCache(endpoint, clusterNameOrID)
+	if isInCache {
+		return clusterNameOrID, nil
+	}
+
+	auxParams := clientWrapper.DefaultAuxiliaryParams()
+	auxParams.ActivityName = listClustersActivityName
+
+	// Get a list of all clusters
+	response, err := clientWrapper.GetClusters(auxParams)
+	if err != nil {
+		switch {
+		case clienterror.IsUnauthorizedError(err):
+			return "", microerror.Mask(errors.NotAuthorizedError)
+
+		case clienterror.IsAccessForbiddenError(err):
+			return "", microerror.Mask(errors.AccessForbiddenError)
+
+		default:
+			return "", microerror.Mask(err)
+		}
+	}
+
+	var (
+		// IDs that correspond to the same cluster name
+		matchingIDs []string
+		// All IDs returned from the response
+		allClusterIDs = make([]string, 0, len(response.Payload))
+	)
+	for _, cluster := range response.Payload {
+		allClusterIDs = append(allClusterIDs, cluster.ID)
+		// Check if this is the cluster we're looking for
+		if matchesValidation(clusterNameOrID, cluster) {
+			matchingIDs = append(matchingIDs, cluster.ID)
+		}
+	}
+
+	CacheIDs(endpoint, allClusterIDs)
+
+	if matchingIDs == nil {
+		// There are no IDs that correspond to that cluster name
+		return "", microerror.Mask(errors.ClusterNotFoundError)
+	} else if len(matchingIDs) > 1 {
+		// There are multiple IDs that correspond to that cluster name
+		// Help the user decide which cluster to pick
+		_, id := handleNameCollision(clusterNameOrID, response.Payload)
```

Comment:
If we don't need the first return value from `handleNameCollision`, I think we should remove it from the functions signature.